File: solutions/189._Rotate_Array/189._Rotate_Array.py
from ast import List


class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        len_nums = len(nums)
        k = k % len_nums
        custom_reverse(nums, len_nums - k, len_nums - 1)
        custom_reverse(nums, 0, len_nums - k - 1)
        custom_reverse(nums, 0, len_nums - 1)


def custom_reverse(nums: List[int], start: int, end: int) -> None:
    while start < end:
        nums[start], nums[end] = nums[end], nums[start]
        start += 1
        end -= 1

File: solutions/medium/189._Rotate_Array/189._Rotate_Array.ts
/**
 Do not return anything, modify nums in-place instead.
 */
function rotate(nums: number[], k: number): void {
  k = k % nums.length;
  reverse(nums, nums.length - k, nums.length - 1);
  reverse(nums, 0, nums.length - k - 1);
  reverse(nums, 0, nums.length - 1);
}

function reverse(nums: number[], start: number, end: number): void {
  while (start < end) {
    const temp = nums[start];
    nums[start] = nums[end];
    nums[end] = temp;
    start++;
    end--;
  }
}

// or this
function shortRotate(nums: number[], k: number): void {
  k = k % nums.length;
  nums.unshift(...nums.splice(nums.length - k, k));
}